File: LeetCode/41. First Missing Positive.py
这道题的思路是看讨论里面一个C++的解法，比我用的遍历用字典记录重复要快一点点，主要思路就是遍历一次，把数字放到对应的位置上，然后重新遍历一次，返回第一个索引与数字不符的即可
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        length = len(nums)
        for index in range(length):
            num = nums[index]
            while (num <= length and num > 0 and nums[index] != nums[num-1]):
                nums[index], nums[num-1] = nums[num-1], nums[index]
                num = nums[index]

        for index in range(length):
            num = nums[index]
            if (index + 1 != num):
                return index + 1
        return length + 1
                

File: LeetCode/41. First Missing Positive.cpp
#include<vector>
using namespace std;

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int l = 0, r = nums.size();
        while(l < r) {
            // 当前位置满足要求
            if (nums[l] == l + 1) {
                ++l;
            }
            // 要么当前位置的数不在 l+1到r这个范围内，要么已经重复了，直接发送到最后垃圾区
            else if (nums[l] <= l || nums[l] > r || nums[nums[l] - 1] == nums[l]) {
                swap(nums[l], nums[--r]);
            }
            // 当前位置的数在l+1到r+1范围内，而且目前该位置还没被同样数字占据，直接发送过去
            else {
                swap(nums[l], nums[nums[l] - 1]);
            }
        }
        return l + 1;
    }
};
